server/async_functions.py

```diff
@@ -0,0 +1,19 @@
+"""
+Some helper functions for common async tasks.
+"""
+import asyncio
+from typing import List
+
+
+async def gather_without_exceptions(
+    tasks: List[asyncio.Task],
+    *exc_classes: type,
+) -> None:
+    """
+    Call gather on a list of tasks, raising the first exception that dosen't
+    match any of the specified exception classes.
+    """
+    results = await asyncio.gather(*tasks, return_exceptions=True)
```

Comment:
From docs
```
If return_exceptions is True, exceptions are treated the same as successful results, and aggregated in the result list.
```
Do we cancel on first exception here?

Reply:
The docs are a little vague, but I've used gather enough to be able to explain how it works. Basically, `gather` always runs every coroutine. Even if one raises an exception, the rest all still get to complete normally. The difference is in how these exceptions are handled. If `return_exceptions` is false, then the call to `gather` will raise the first exception that it encounters. If `return_exceptions` is true, then all exceptions will be aggregated into the result list so you can iterate through them later.